core/src/main/java/org/elasticsearch/search/aggregations/metrics/percentiles/ParsedPercentiles.java

```diff
@@ -0,0 +1,168 @@
+                return iterator.hasNext();
+            }
+
+            @Override
+            public Percentile next() {
+                Map.Entry<Double, Double> next = iterator.next();
+                return new InternalPercentile(next.getKey(), next.getValue());
+            }
+        };
+    }
+
+    @Override
+    protected XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        if (keyed) {
+            builder.startObject(InternalAggregation.CommonFields.VALUES.getPreferredName());
+            for (Map.Entry<Double, Double> percentile : percentiles.entrySet()) {
+                Double key = percentile.getKey();
+                builder.field(String.valueOf(key), percentile.getValue());
+
+                String valueAsString = percentilesAsString.get(key);
+                if (valueAsString != null && valueAsString.isEmpty() == false) {
+                    builder.field(key + "_as_string", valueAsString);
+                }
+            }
+            builder.endObject();
+        } else {
+            builder.startArray(InternalAggregation.CommonFields.VALUES.getPreferredName());
+            for (Map.Entry<Double, Double> percentile : percentiles.entrySet()) {
+                Double key = percentile.getKey();
+                builder.startObject();
+                {
+                    builder.field(InternalAggregation.CommonFields.KEY.getPreferredName(), key);
+                    builder.field(InternalAggregation.CommonFields.VALUE.getPreferredName(), percentile.getValue());
+                    String valueAsString = percentilesAsString.get(key);
+                    if (valueAsString != null && valueAsString.isEmpty() == false) {
+                        builder.field(InternalAggregation.CommonFields.VALUE_AS_STRING.getPreferredName(), valueAsString);
+                    }
+                }
+                builder.endObject();
+            }
+            builder.endArray();
+        }
+        return builder;
+    }
+
+    protected static void declarePercentilesFields(ObjectParser<? extends ParsedPercentiles, Void> objectParser) {
+        ParsedAggregation.declareCommonFields(objectParser);
+
+        objectParser.declareField((parser, aggregation, context) -> {
+            XContentParser.Token token = parser.currentToken();
+            if (token == XContentParser.Token.START_OBJECT) {
+                aggregation.keyed = true;
+                while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                    if (token.isValue()) {
+                        if (token == XContentParser.Token.VALUE_NUMBER) {
+                            aggregation.addPercentile(Double.valueOf(parser.currentName()), parser.doubleValue());
+
+                        } else if (token == XContentParser.Token.VALUE_STRING) {
+                            int i = parser.currentName().indexOf("_as_string");
+                            if (i > 0) {
```

Comment:
The answer is: the format is not settable through the request, but the default formatter varies depending on the field type. With dates, the key_as_string will always be printed out then.